app/contracts/members/base_contract.rb

```diff
@@ -50,6 +52,12 @@ def assignable_projects
 
     private
 
+    def validate_no_global_placeholder
+      return unless principal.is_a?(PlaceholderUser)
+
+      errors.add :principal, :invalid if project_id.nil?
+    end
+
```

Comment:
Even without being a `PlaceholderUser`, validation would fail because `project_id` is blank.

Prying into the spec you added, the contract fails with the following errors: `{roles: ["has an unassignable role."], project: ["can't be blank."], principal: ["is invalid."]}` (I was curious to see the error message).

So it can't hurt, but I'm not sure that's actually useful.

Reply:
But pushing it further, I actually see that this "project can't be blank" error is not added if current user is admin.
So only admins can set global roles?

So to make the unit test more valid, the current user should be an admin, right? so the other errors about roles and project do not trigger.

I'll continue reviewing tomorrow.